File: 289.GameofLife.py
from typing import List

class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        m, n = len(board), len(board[0])
        neighbors = [[0, 1], [1, 1], [1, 0], [1, -1], [0, -1], [-1, -1], [-1, 0], [-1, 1]]
        for i in range(m):
            for j in range(n):
                temp = [board[i + dx][j + dy] for dx, dy in neighbors if 0 <= i + dx < m and 0 <= j + dy < n]
                print(temp)
                lives = temp.count(1) + temp.count(2)
                if board[i][j]:
                    if lives < 2 or lives > 3:
                        board[i][j] = 2
                    else:
                        board[i][j] = 1

                else:
                    if lives == 3:
                        board[i][j] = 3
                    else:
                        board[i][j] = 0
        
        for i in range(m):
            for j in range(n):
                if board[i][j] == 2:
                    board[i][j] = 0
                elif board[i][j] == 3:
                    board[i][j] = 1
        

a = Solution()
a.gameOfLife([[0,1,0],[0,0,1],[1,1,1],[0,0,0]])

File: Medium/289.Game_of_Life.java
package Medium;

class Solution {
    public void gameOfLife(int[][] board) {
        int m = board.length, n = board[0].length;
        int[][] movements = { {0, 1}, {0, -1}, {1, 0}, {-1, 0}, 
                              {1, 1}, {1, -1}, {-1, 1}, {-1, -1} };
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int count = 0;
                for (int[] movement : movements) {
                    int x = i + movement[0], y = j + movement[1];
                    if (0 <= x && x < m && 0 <= y && y < n) {
                        if (board[x][y] == 1 || board[x][y] == -1) count++;
                    }
                }
                if (count < 2) {
                    if (board[i][j] == 1) board[i][j] = -1;
                } else if (count == 3) {
                    if (board[i][j] == 0) board[i][j] = -2;
                } else if (4 <= count) {
                    if (board[i][j] == 1) board[i][j] = -1;
                }
            }
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == -1) board[i][j] = 0;
                else if (board[i][j] == -2) board[i][j] = 1;
            }
        }
    }
}